oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+
+We expect the following packages - which are used in the vast majority of MFEs today - could all be shared between MFEs.
+
+.. list-table::
+   :widths: 30 20 50
+
+   * - **Package**
+     - **Estimated Size**
+     - **Notes**
+   * - @edx/brand
+     - Variable
+     -
+   * - @edx/frontend-component-footer
+     - 88.1k
+     -
+   * - @edx/frontend-component-header
+     - 156.9k
+     -
+   * - @edx/frontend-platform
+     - 355.3k
+     -
+   * - @openedx/paragon (or @edx/paragon)
+     - ~950k
+     - bundlephobia.com threw an error on the Paragon build, but we expect it’s the largest package here.
+   * - classnames
+     - 0.8k
+     -
+   * - core-js
+     - 241.1k
+     -
+   * - prop-types
+     - 0.9k
+     -
+   * - react
+     - 6.4k
+     -
+   * - react-dom
+     - 130.2k
+     -
+   * - react-redux
+     - 11.2k
+     -
+   * - react-router
+     - 58.9k
+     -
+   * - react-router-dom
+     - 77.1k
+     -
+   * - redux
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit. This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~2,087.2k unzipped
+
+Sharing these dependencies has several important positive effects:
+
+- Enforces runtime consistency across all types of dependencies
+- Reduces runtime bundle size
+
+Note that "build time" and "dependency maintenance" are not mentioned above.
+
+Interestingly, because of the way Webpack module federation's dependency sharing is implemented, this may not have a large impact on build time or dependency maintenance. "Guest" MFEs still need to have the dependencies installed for the build to succeed, and artifacts for them will still be created and deployed - they just won't generally be used at runtime unless there's a version conflict.
+
+These dependencies still need to exist in the package.json file (and be installed by npm) for the build to succeed, and for IDE hinting, auto completion, linting, etc.
+
+Maintaining Dependency Consistency
+==================================
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity.
+
+We expect that this may need to take a number of possible forms:
+
+Process
+-------
+
+We need to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent. We also need a process to migrate through breaking changes in third-party dependencies.
+
+Best Practices
+--------------
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon).
```

Comment:
Paragon is a tough one... it's still pretty immature in some places and there's definitely going to be a tradeoff here between stability and development velocity.

Reply:
One approach we tried (or maybe only considered?) with Paragon once upon a time was to just create new versions of components when a breaking change is strictly necessary, i.e., you end up with `ButtonV2`.  We could retire deprecated components with the DEPR process aligned to platform releases or something, rather than having random major version bumps whenever it's convenient.

One way or another, I think we need to get really strict about breaking changes in our libraries, like, paragon and frontend-platform, or this gets way harder.

Reply:
I've added some specific recommendations around maintaining dependency consistency, akin to what I wrote here.